src/ty.md

```diff
@@ -17,110 +179,378 @@ tcx: TyCtxt<'tcx>
+then we would first look at the overall type. We would see that there are no substitutions to be
+made at the outer level, so we would descend one level and look at `Adt(Vec, &[Param(X)])`. There
+are still no substitutions to be made here, so we would descend again. Now we are looking at
+`Param(X)`, which can be substituted, so we replace it with `u32`. We can’t descend any more, so we
+are done, and  the overall result is `Adt(Vec, &[Adt(Vec, &[u32])])`.
+
+One last thing to mention: often when folding over a `TypeFoldable`, we don’t want to change most
+things. We only want to do something when we reach a type. That means there may be a lot of
+`TypeFoldable` types whose implementations basically just forward to their fields’ `TypeFoldable`
+implementations. Such implementations of `TypeFoldable` tend to be pretty tedious to write by hand.
+For this reason, there is a `derive` macro that allows you to `#![derive(TypeFoldable)]`. It is
+defined
+[here](https://github.com/rust-lang/rust/blob/master/src/librustc_macros/src/type_foldable.rs).
+
+**`subst`** In the case of substitutions the [actual
+folder](https://github.com/rust-lang/rust/blob/04e69e4f4234beb4f12cc76dcc53e2cc4247a9be/src/librustc/ty/subst.rs#L467-L482)
+is going to be doing the indexing we’ve already mentioned. There we define a `Folder` and call
+`fold_with` on the `TypeFoldable` to process yourself.  Then
+[fold_ty](https://github.com/rust-lang/rust/blob/04e69e4f4234beb4f12cc76dcc53e2cc4247a9be/src/librustc/ty/subst.rs#L545-L573)
+the method that process each type it looks for a `ty::Param` and for those it replaces it for
+something from the list of substitutions, otherwise recursively process the type.  To replace it,
+calls
+[ty_for_param](https://github.com/rust-lang/rust/blob/04e69e4f4234beb4f12cc76dcc53e2cc4247a9be/src/librustc/ty/subst.rs#L589-L624)
+and all that does is index into the list of substitutions with the index of the `Param`.
+
+## Errors
+
+It is possible for the indices in `Param` to not match with what we expect. For example, the index
+could be out of bounds or it could be the index of a lifetime when we were expecting a type. These
+sorts of errors would be caught earlier in the compiler when translating from a `hir::Ty` to a
+`ty::Ty`. If they occur later, that is a compiler bug.
+
+There is a `TyKind::Error` that is produced when the user makes this sort of error. The idea is that
+we would propagate this type and suppress other errors that come up due to it so as not to overwhelm
+the user with cascading compiler error messages.
+
+## Question: Why not substitute “inside” the adt-def?
+
+The adt-def logically represents just the *name* of the struct (e.g., `Vec`). Its contents are
+always assumed to be expressed in terms of the “internal” or “self-view” — i.e., expressed in terms
+of the generics on the struct. If nothing else, this has an efficiency win:
```

Comment:
Even though this is reflecting what I said, I found it kind of confusing. =) Basically, we're arguing for why we represent `MyStruct<u32>` as the pair of an "unsubstituted" definition (`MyStruct`, the `AdtDef`) combined with a list of substitutions (the `[u32]`). This requires us then to apply those substitutions whenever we extract a type from the struct definition, such as the type of a field.

The alternative would be to do the substitution "eagerly" -- i.e., create a new, fully substituted form of the `AdtDef`, where the types of all fields are already substituted. I think it's worth spelling out this alternative in the text to help explain what you are arguing *against*.